util/file.py

```diff
@@ -516,3 +516,40 @@ def slurp_file(fname, maxSizeMb=50):
     with open_or_gzopen(fname) as f:
         return f.read()
 
+def is_broken_link(filename):
+    # isfile() returns True if a file, or a working link
+    if os.path.isfile(filename) or os.path.isdir(filename):
+        return False
+    # otherwise if this is a link
+    if os.path.islink(filename):
+        # os.path.exists() returns false in the case of broken symlinks
+        return not os.path.exists(filename)
+    return False
+
+
+def remove_broken_symlinks(rootdir, followlinks=False):
+    """
+        This function removes broken symlinks within a directory, 
+        doing the same in each child directory as well (though not following
+        functional symlinks, unless they're directories and followlinks=True).
+        @param followlinks: only applies to directory links as per os.walk
+    """
+
+    broken_links_to_remove = []
+
+    # first check to see if the input is itself a broken link
+    if is_broken_link(rootdir):
+        log.info("Removing broken symlink: %s", rootdir)
+        broken_links_to_remove.append(rootdir.rstrip("/"))
+    else:
+        # otherwise traverse the directory hierarchy
+        log.info("Checking for broken symlinks to remove in: %s", rootdir)
+        for rootpath, subfolders, files in os.walk(rootdir, followlinks=followlinks):
+            for filename in files:
+                fpath = os.path.join(rootpath, filename)
+                if is_broken_link(fpath):
+                    broken_links_to_remove.append(fpath.rstrip("/"))
+
+    for lpath in broken_links_to_remove:
+        log.info("Removing broken symlink: %s", lpath)
```

Comment:
Since this is in `util.file`, I wonder if it might be better if the function was silent (from a logging point of view) and it just returned the list `broken_links_to_remove` as the return value. The higher up caller can then quickly interpret the the result as a boolean (did it do anything) or count the values or `log.info` the returned values if the verbosity is desired.